Source/ACE/Network/GameMessages/Messages/GameMessageFellowshipFullUpdate.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using ACE.Common.Extensions;
+using ACE.Entity;
+
+namespace ACE.Network.GameMessages.Messages
+{
+    public class GameMessageFellowshipFullUpdate : GameMessage
+    {
+        public GameMessageFellowshipFullUpdate(Session session, Fellowship fellowship)
+            : base(GameMessageOpcode.GameEvent, GameMessageGroup.Group09)
+        {
+            Writer.Write(session.Player.Guid.Full);
+            Writer.Write(session.GameEventSequence++);
+            Writer.Write((uint)GameEvent.GameEventType.FellowshipFullUpdate);
+
+            // the current number of fellowship members
+            Writer.Write((UInt16)fellowship.FellowshipMembers.Count);
+
+            // ????
+            Writer.Write((byte)0x10);
+            Writer.Write((byte)0x00);
+
+            // --- FellowInfo ---
+            foreach (Player fellow in fellowship.FellowshipMembers)
+            {
+                Writer.Write(fellow.Guid.Full);
```

Comment:
This part is actually tricky :-/.

We have a multi-threaded server, and you can't just go reading the Player's values without any form of concurrency control.

The way we manage concurrency control between in-game entities is with an event-based system we call "Actions" the most common way we use "Actions" is with an "ActionChain" (or a series of actions which happen one after the other).

In summary:
Actions run on a WorldObject's turn (a Player is a WorldObject).  This guarantees that any accesses to that object's data wont race with any others.

The proper code skeleton to do this would look something like this:

ActionChain chain = new ActionChain();
foreach (Player fellow in fellowhsip.FellowshipMembers)
{
    chain.AddAction(player, () =>
    {
        ...Write Player properties here...
    }
}

chain.AddAction(session.Player, ()
{
    ... Write fellowship name and leader guid here ...
}
chain.EnqueueChain();